solidity/contracts/Rewards.sol

```diff
@@ -0,0 +1,518 @@
+    using SafeERC20 for IERC20;
+
+    IERC20 token;
+
+    // Total number of keep tokens to distribute.
+    uint256 totalRewards;
+    // Rewards that haven't been allocated to finished intervals
+    uint256 unallocatedRewards;
+    // Rewards that have been paid out;
+    // `token.balanceOf(address(this))` should always equal
+    // `totalRewards.sub(paidOutRewards)`
+    uint256 paidOutRewards;
+
+    // Length of one interval in seconds (timestamp diff).
+    uint256 termLength;
+    // Timestamp of first interval beginning.
+    uint256 firstIntervalStart;
+
+    // Minimum number of keep submissions for each interval.
+    uint256 minimumKeepsPerInterval;
+
+    // Array representing the percentage of unallocated rewards
+    // available for each reward interval.
+    uint256[] intervalWeights; // percent array
+    // Mapping of interval number to tokens allocated for the interval.
+    uint256[] intervalAllocations;
+
+    // mapping of keeps to booleans.
+    // True if the keep has been used to claim a reward.
+    mapping(bytes32 => bool) claimed;
+
+    // Mapping of interval to number of keeps created in/before the interval
+    mapping(uint256 => uint256) keepsByInterval;
+
+    // Mapping of interval to number of keeps whose rewards have been paid out,
+    // or reallocated because the keep closed unhappily
+    mapping(uint256 => uint256) intervalKeepsProcessed;
+
+    constructor (
+        uint256 _termLength,
+        address _token,
+        uint256 _minimumKeepsPerInterval,
+        uint256 _firstIntervalStart,
+        uint256[] memory _intervalWeights
+    ) public {
+        token = IERC20(_token);
+        termLength = _termLength;
+        firstIntervalStart = _firstIntervalStart;
+        minimumKeepsPerInterval = _minimumKeepsPerInterval;
+        intervalWeights = _intervalWeights;
+    }
+
+    /// @notice `approveAndCall` funds the rewards contract.
+    /// @dev Adds the received amount of tokens
+    /// to `totalRewards` and `unallocatedRewards`.
+    /// May be called at any time, even after allocating some intervals.
+    /// Changes to `unallocatedRewards`
+    /// will take effect on subsequent interval allocations.
+    /// If the reward contract has received tokens outside `approveAndCall`,
+    /// this collects them as well.
+    function receiveApproval(
+        address _from,
+        uint256 _value,
+        address _token,
+        bytes memory
+    ) public {
+        require(IERC20(_token) == token, "Unsupported token");
+
+        token.safeTransferFrom(_from, address(this), _value);
+
+        uint256 currentBalance = token.balanceOf(address(this));
+        uint256 beforeBalance = _unpaidRewards();
+        require(
+            currentBalance >= beforeBalance,
+            "Reward contract has lost tokens"
+        );
+
+        uint256 addedBalance = currentBalance.sub(beforeBalance);
+
+        totalRewards += addedBalance;
+        unallocatedRewards += addedBalance;
```

Comment:
Good call. I think there are justified cases for using unsafe math in tightly optimized loops where correctness is reliably maintained otherwise, but this is definitely not one of them.